File: csharp/bob/Bob.cs
using System;
using System.Text.RegularExpressions;

public static class Bob
{
    public static string Response(string input)
    {
        string statement = input.Trim();

        if (String.IsNullOrEmpty(statement)) 
            return "Fine. Be that way!";

        if (IsShouting(statement) && IsQuestion(statement))
            return "Calm down, I know what I'm doing!";

        if (IsShouting(statement)) 
            return "Whoa, chill out!";

        if (IsQuestion(statement))
            return "Sure.";

        return "Whatever.";
    }

    private static bool IsQuestion(string input) => 
        input.EndsWith("?");

    private static bool IsShouting(string input)
    {
        char[] letters = Regex.Replace(input, @"[^a-zA-Z]", "").ToCharArray();

        return letters.Length > 0 
            && Array.TrueForAll(letters, Char.IsUpper);
    }
}

File: fsharp/bob/Bob.fs
module Bob

open System

let charCollector (c: char): string = 
    if Char.IsLetter c
    then Char.ToString c
    else ""

let getLetters (statement: string): string = 
    String.collect charCollector statement

let (|Nothing|_|) (statement: string) = 
    if statement.Length = 0 
    then Some Nothing
    else None

let (|Shouting|_|) (statement: string) = 
    let letters = getLetters statement
    if letters.Length > 0 && letters.ToUpper() = letters
    then Some Shouting
    else None

let (|Question|_|) (statement: string) =
    if statement.EndsWith("?") 
    then Some Question
    else None

let response (input: string) = 
    match input.Trim() with
    | Nothing -> "Fine. Be that way!"
    | Shouting & Question -> "Calm down, I know what I'm doing!"
    | Shouting -> "Whoa, chill out!"
    | Question -> "Sure."
    | _ -> "Whatever."